File: sdk/node/examples/basic/test_generate_object.ts
#!/usr/bin/env npx tsx
/**
 * Integration test: generate_object tool via A3S Code Node.js SDK
 *
 * Tests structured JSON output using MiniMax-M2.7-highspeed model.
 * Config is loaded from .a3s/config.acl (never hardcoded).
 *
 * Run: cd crates/code/sdk/node/examples && npx tsx basic/test_generate_object.ts
 */

import { Agent, Session } from '../../index.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ============================================================================
// Config loader — reads .a3s/config.acl at runtime, never hardcodes secrets
// ============================================================================

function loadConfig(): string {
  const candidates = [
    path.resolve(__dirname, '../../../../../../.a3s/config.acl'),
    path.resolve(__dirname, '../../../../../.a3s/config.acl'),
    path.resolve(__dirname, '../../../../.a3s/config.acl'),
  ];
  for (const p of candidates) {
    if (fs.existsSync(p)) {
      return fs.readFileSync(p, 'utf-8');
    }
  }
  throw new Error(`.a3s/config.acl not found. Tried: ${candidates.join(', ')}`);
}

// ============================================================================
// Test runner
// ============================================================================

interface TestResult {
  name: string;
  passed: boolean;
  error?: string;
  durationMs: number;
}

const results: TestResult[] = [];

async function runTest(name: string, fn: () => Promise<void>): Promise<void> {
  const start = Date.now();
  try {
    await fn();
    results.push({ name, passed: true, durationMs: Date.now() - start });
    console.log(`  ✓ PASS (${Date.now() - start}ms): ${name}`);
  } catch (e) {
    results.push({ name, passed: false, error: String(e), durationMs: Date.now() - start });
    console.log(`  ✗ FAIL (${Date.now() - start}ms): ${name}`);
    console.log(`    Error: ${e}`);
  }
}

// ============================================================================
// Tests
// ============================================================================

async function main() {
  console.log('='.repeat(70));
  console.log('generate_object Integration Test (Node.js SDK + MiniMax)');
  console.log('='.repeat(70));

  const configContent = loadConfig();
  const workspace = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-genobj-test-'));
  console.log(`Workspace: ${workspace}\n`);

  const agent = await Agent.create(configContent);
  const session = agent.session(workspace, {
    permissionPolicy: { defaultDecision: 'allow' },
  });

  // ── Test 1: Direct tool call — simple extraction ──────────────────────────

  await runTest('generate_object: extract person info (tool mode)', async () => {
    const result = await session.tool('generate_object', {
      schema: {
        type: 'object',
        required: ['name', 'age', 'city'],
        properties: {
          name: { type: 'string' },
          age: { type: 'integer' },
          city: { type: 'string' },
        },
      },
      prompt: 'Extract: "Alice Chen is 28 years old and lives in Shanghai."',
      schema_name: 'person',
      mode: 'tool',
    });

    if (result.exitCode !== 0) throw new Error(`Tool failed: ${result.output}`);
    const parsed = JSON.parse(result.output);
    const obj = parsed.object;
    if (obj.name.toLowerCase() !== 'alice chen') throw new Error(`name: ${obj.name}`);
    if (obj.age !== 28) throw new Error(`age: ${obj.age}`);
    if (!obj.city.toLowerCase().includes('shanghai')) throw new Error(`city: ${obj.city}`);
    console.log(`    Result: ${JSON.stringify(obj)}`);
  });

  // ── Test 2: Enum classification ───────────────────────────────────────────

  await runTest('generate_object: sentiment classification with enum', async () => {
    const result = await session.tool('generate_object', {
      schema: {
        type: 'object',
        required: ['sentiment', 'confidence'],
        properties: {
          sentiment: { type: 'string', enum: ['positive', 'negative', 'neutral'] },
          confidence: { type: 'number', minimum: 0, maximum: 1 },
        },
      },
      prompt: 'Classify sentiment: "This is the worst product I have ever used. Total waste of money."',
      schema_name: 'sentiment',
    });

    if (result.exitCode !== 0) throw new Error(`Tool failed: ${result.output}`);
    const parsed = JSON.parse(result.output);
    const obj = parsed.object;
    if (obj.sentiment !== 'negative') throw new Error(`sentiment: ${obj.sentiment}`);
    if (obj.confidence < 0 || obj.confidence > 1) throw new Error(`confidence: ${obj.confidence}`);
    console.log(`    Result: sentiment=${obj.sentiment}, confidence=${obj.confidence}`);
  });

  // ── Test 3: Nested schema ─────────────────────────────────────────────────

  await runTest('generate_object: nested schema (API error)', async () => {
    const result = await session.tool('generate_object', {
      schema: {
        type: 'object',
        required: ['error'],
        properties: {
          error: {
            type: 'object',
            required: ['code', 'message', 'type'],
            properties: {
              code: { type: 'integer' },
              message: { type: 'string', minLength: 3 },
              type: { type: 'string', enum: ['not_found', 'unauthorized', 'validation'] },
            },
          },
        },
      },
      prompt: 'Generate a 404 not_found API error for missing user resource.',
      schema_name: 'api_error',
    });

    if (result.exitCode !== 0) throw new Error(`Tool failed: ${result.output}`);
    const parsed = JSON.parse(result.output);
    const obj = parsed.object;
    if (obj.error.code !== 404) throw new Error(`code: ${obj.error.code}`);
    if (obj.error.type !== 'not_found') throw new Error(`type: ${obj.error.type}`);
    if (obj.error.message.length < 3) throw new Error(`message too short`);
    console.log(`    Result: ${JSON.stringify(obj.error)}`);
  });

  // ── Test 4: Array output ──────────────────────────────────────────────────

  await runTest('generate_object: array of items', async () => {
    const result = await session.tool('generate_object', {
      schema: {
        type: 'object',
        required: ['items'],
        properties: {
          items: {
            type: 'array',
            minItems: 3,
            maxItems: 5,
            items: {
              type: 'object',
              required: ['name', 'category'],
              properties: {
                name: { type: 'string' },
                category: { type: 'string', enum: ['fruit', 'vegetable', 'grain'] },
              },
            },
          },
        },
      },
      prompt: 'List 3 food items with their categories.',
      schema_name: 'food_list',
    });

    if (result.exitCode !== 0) throw new Error(`Tool failed: ${result.output}`);
    const parsed = JSON.parse(result.output);
    const items = parsed.object.items;
    if (!Array.isArray(items)) throw new Error('items is not array');
    if (items.length < 3 || items.length > 5) throw new Error(`items count: ${items.length}`);
    for (const item of items) {
      if (!['fruit', 'vegetable', 'grain'].includes(item.category)) {
        throw new Error(`invalid category: ${item.category}`);
      }
    }
    console.log(`    Result: ${items.length} items — ${items.map((i: any) => i.name).join(', ')}`);
  });

  // ── Test 5: Streaming via session.stream (agent decides to use tool) ──────

  await runTest('generate_object: streaming via session.send (agent-driven)', async () => {
    const result = await session.send(
      'Use the generate_object tool to extract the following into a structured object with fields "title" (string), "year" (integer), "genre" (string): The movie "Inception" was released in 2010 and is a sci-fi thriller.'
    );
    // The agent should have called generate_object and returned the result
    if (!result.text.toLowerCase().includes('inception') && !result.text.includes('"title"')) {
      // Check if it used the tool
      if (result.toolCallsCount === 0) {
        throw new Error(`Agent did not use generate_object tool. Response: ${result.text.slice(0, 200)}`);
      }
    }
    console.log(`    Tool calls: ${result.toolCallsCount}, tokens: ${result.totalTokens}`);
  });

  // ── Test 6: Prompt mode ───────────────────────────────────────────────────

  await runTest('generate_object: prompt mode fallback', async () => {
    const result = await session.tool('generate_object', {
      schema: {
        type: 'object',
        required: ['answer'],
        properties: {
          answer: { type: 'integer' },
        },
      },
      prompt: 'What is 7 * 8? Return the answer.',
      schema_name: 'math',
      mode: 'prompt',
    });

    if (result.exitCode !== 0) throw new Error(`Tool failed: ${result.output}`);
    const parsed = JSON.parse(result.output);
    if (parsed.object.answer !== 56) throw new Error(`answer: ${parsed.object.answer}`);
    console.log(`    Result: 7*8 = ${parsed.object.answer}`);
  });

  // ── Summary ───────────────────────────────────────────────────────────────

  console.log('\n' + '='.repeat(70));
  const passed = results.filter(r => r.passed).length;
  const failed = results.filter(r => !r.passed).length;
  console.log(`Results: ${passed} passed, ${failed} failed, ${results.length} total`);
  console.log('='.repeat(70));

  // Cleanup
  fs.rmSync(workspace, { recursive: true, force: true });

  if (failed > 0) process.exit(1);
}

main().catch(e => {
  console.error('Fatal:', e);
  process.exit(1);
});

File: sdk/python/examples/test_generate_object.py
#!/usr/bin/env python3
"""
Integration test: generate_object tool via A3S Code Python SDK

Tests structured JSON output using MiniMax-M2.7-highspeed model.
Config is loaded from .a3s/config.acl at runtime (never hardcoded).

Run: cd crates/code/sdk/python && python examples/test_generate_object.py
"""

import json
import os
import sys
import tempfile
import time
from pathlib import Path


def load_config() -> str:
    """Load .a3s/config.acl from the repo root."""
    script_dir = Path(__file__).resolve().parent
    candidates = [
        script_dir / "../../../../../../.a3s/config.acl",
        script_dir / "../../../../../.a3s/config.acl",
        script_dir / "../../../../.a3s/config.acl",
        script_dir / "../../../.a3s/config.acl",
    ]
    for p in candidates:
        resolved = p.resolve()
        if resolved.exists():
            return resolved.read_text()
    raise FileNotFoundError(".a3s/config.acl not found. Cannot run integration test.")


# ============================================================================
# Test runner
# ============================================================================

results: list[dict] = []


def run_test(name: str, fn):
    start = time.time()
    try:
        fn()
        elapsed = int((time.time() - start) * 1000)
        results.append({"name": name, "passed": True, "ms": elapsed})
        print(f"  ✓ PASS ({elapsed}ms): {name}")
    except Exception as e:
        elapsed = int((time.time() - start) * 1000)
        results.append({"name": name, "passed": False, "error": str(e), "ms": elapsed})
        print(f"  ✗ FAIL ({elapsed}ms): {name}")
        print(f"    Error: {e}")


# ============================================================================
# Tests
# ============================================================================

def main():
    print("=" * 70)
    print("generate_object Integration Test (Python SDK + MiniMax)")
    print("=" * 70)

    from a3s_code import Agent, SessionOptions, PermissionPolicy

    config_content = load_config()
    workspace = tempfile.mkdtemp(prefix="a3s-genobj-py-")
    print(f"Workspace: {workspace}\n")

    agent = Agent.create(config_content)
    policy = PermissionPolicy(default_decision="allow")
    opts = SessionOptions()
    opts.permission_policy = policy
    session = agent.session(workspace, opts)

    # ── Test 1: Simple extraction ────────────────────────────────────────────

    def test_extract_person():
        result = session.tool("generate_object", {
            "schema": {
                "type": "object",
                "required": ["name", "age", "city"],
                "properties": {
                    "name": {"type": "string"},
                    "age": {"type": "integer"},
                    "city": {"type": "string"},
                },
            },
            "prompt": 'Extract: "Bob Zhang is 32 years old and lives in Beijing."',
            "schema_name": "person",
            "mode": "tool",
        })
        assert result.exit_code == 0, f"Tool failed: {result.output}"
        parsed = json.loads(result.output)
        obj = parsed["object"]
        assert obj["name"].lower() == "bob zhang", f"name: {obj['name']}"
        assert obj["age"] == 32, f"age: {obj['age']}"
        assert "beijing" in obj["city"].lower(), f"city: {obj['city']}"
        print(f"    Result: {json.dumps(obj)}")

    run_test("generate_object: extract person info", test_extract_person)

    # ── Test 2: Enum classification ──────────────────────────────────────────

    def test_sentiment():
        result = session.tool("generate_object", {
            "schema": {
                "type": "object",
                "required": ["sentiment"],
                "properties": {
                    "sentiment": {"type": "string", "enum": ["positive", "negative", "neutral"]},
                },
            },
            "prompt": 'Classify: "I absolutely love this new feature, it works perfectly!"',
            "schema_name": "sentiment",
        })
        assert result.exit_code == 0, f"Tool failed: {result.output}"
        parsed = json.loads(result.output)
        obj = parsed["object"]
        assert obj["sentiment"] == "positive", f"sentiment: {obj['sentiment']}"
        print(f"    Result: sentiment={obj['sentiment']}")

    run_test("generate_object: sentiment classification", test_sentiment)

    # ── Test 3: Nested schema ────────────────────────────────────────────────

    def test_nested():
        result = session.tool("generate_object", {
            "schema": {
                "type": "object",
                "required": ["book"],
                "properties": {
                    "book": {
                        "type": "object",
                        "required": ["title", "author", "year", "genres"],
                        "properties": {
                            "title": {"type": "string"},
                            "author": {"type": "string"},
                            "year": {"type": "integer", "minimum": 1800, "maximum": 2030},
                            "genres": {
                                "type": "array",
                                "items": {"type": "string"},
                                "minItems": 1,
                            },
                        },
                    }
                },
            },
            "prompt": 'Extract book info: "1984 by George Orwell, published 1949, genres: dystopian fiction, political fiction"',
            "schema_name": "book_info",
        })
        assert result.exit_code == 0, f"Tool failed: {result.output}"
        parsed = json.loads(result.output)
        obj = parsed["object"]["book"]
        assert "1984" in obj["title"], f"title: {obj['title']}"
        assert "orwell" in obj["author"].lower(), f"author: {obj['author']}"
        assert obj["year"] == 1949, f"year: {obj['year']}"
        assert len(obj["genres"]) >= 1, f"genres: {obj['genres']}"
        print(f"    Result: {obj['title']} by {obj['author']} ({obj['year']})")

    run_test("generate_object: nested schema (book)", test_nested)

    # ── Test 4: Array output ─────────────────────────────────────────────────

    def test_array():
        result = session.tool("generate_object", {
            "schema": {
                "type": "object",
                "required": ["colors"],
                "properties": {
                    "colors": {
                        "type": "array",
                        "minItems": 3,
                        "maxItems": 3,
                        "items": {
                            "type": "object",
                            "required": ["name", "hex"],
                            "properties": {
                                "name": {"type": "string"},
                                "hex": {"type": "string", "pattern": "^#[0-9a-fA-F]{6}$"},
                            },
                        },
                    }
                },
            },
            "prompt": "List exactly 3 colors with their hex codes: red, green, blue.",
            "schema_name": "colors",
        })
        assert result.exit_code == 0, f"Tool failed: {result.output}"
        parsed = json.loads(result.output)
        colors = parsed["object"]["colors"]
        assert len(colors) == 3, f"Expected 3 colors, got {len(colors)}"
        for c in colors:
            assert c["hex"].startswith("#"), f"Invalid hex: {c['hex']}"
            assert len(c["hex"]) == 7, f"Hex wrong length: {c['hex']}"
        print(f"    Result: {', '.join(c['name'] + '=' + c['hex'] for c in colors)}")

    run_test("generate_object: array with pattern validation", test_array)

    # ── Test 5: Prompt mode ──────────────────────────────────────────────────

    def test_prompt_mode():
        result = session.tool("generate_object", {
            "schema": {
                "type": "object",
                "required": ["result"],
                "properties": {
                    "result": {"type": "integer"},
                },
            },
            "prompt": "What is 13 * 7? Return the numeric result.",
            "schema_name": "math",
            "mode": "prompt",
        })
        assert result.exit_code == 0, f"Tool failed: {result.output}"
        parsed = json.loads(result.output)
        assert parsed["object"]["result"] == 91, f"Expected 91, got {parsed['object']['result']}"
        print(f"    Result: 13*7 = {parsed['object']['result']}")

    run_test("generate_object: prompt mode (math)", test_prompt_mode)

    # ── Test 6: Agent-driven (session.send) ──────────────────────────────────

    def test_agent_driven():
        result = session.send(
            'Use the generate_object tool to produce a JSON object with schema '
            '{"type":"object","required":["language","paradigm"],"properties":'
            '{"language":{"type":"string"},"paradigm":{"type":"string"}}} '
            'for the following: "Rust is a systems programming language with a focus on safety."'
        )
        assert result.text is not None
        # Agent should have used the tool
        print(f"    Tool calls: {result.tool_calls_count}, tokens: {result.total_tokens}")

    run_test("generate_object: agent-driven via session.send", test_agent_driven)

    # ── Summary ──────────────────────────────────────────────────────────────

    print("\n" + "=" * 70)
    passed = sum(1 for r in results if r["passed"])
    failed = sum(1 for r in results if not r["passed"])
    print(f"Results: {passed} passed, {failed} failed, {len(results)} total")
    print("=" * 70)

    # Cleanup
    import shutil
    shutil.rmtree(workspace, ignore_errors=True)

    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    main()
